text/0000-try-expr.md

```diff
@@ -0,0 +1,886 @@
+    // Note: OK-wrapping is assumed here, but it is not the goal of this RFC
+    // to decide either in favor or against OK-wrapping.
+    x + y
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+[list of keywords]: https://doc.rust-lang.org/book/second-edition/appendix-01-keywords.html#keywords-currently-in-use
+
+The word `try` is reserved as a keyword in the [list of keywords]
+in Rust edition 2018 and later editions.
+
+The keyword `try` is used in "try expressions" of the form `try { .. }`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are two main drawbacks to the `try` keyword.
+
+## Association with exception handling - Both a pro and con
+
+> I think that there is a belief – one that I have shared from time to time – that it is not helpful to use familiar keywords unless the semantics are a perfect match, the concern being that they will setup an intuition that will lead people astray. I think that is a danger, but it works both ways: those intuitions also help people to understand, particularly in the early days. So it’s a question of “how far along will you get before the differences start to matter” and “how damaging is it if you misunderstand for a while”.
+>
+> [..]
+>
+> Rust has a lot of concepts to learn. If we are going to succeed, it’s essential that people can learn them a bit at a time, and that we not throw everything at you at once. I think we should always be on the lookout for places where we can build on intuitions from other languages; it doesn’t have to be a 100% match to be useful.
+
+\- [Niko Matsakis](https://internals.rust-lang.org/t/bikeshed-rename-catch-blocks-to-fallible-blocks/7121/4)
+
+For some people, the association to `try { .. } catch { .. }` in languages such
+as Java, and others in the [prior-art] section, is unhelpful wrt. teachability
+because they see the explicit, reifed, and manually propagated exceptions in
+Rust as something very different than the much more implicit exception handling
+stories in Java et al.
+
+[`ExceptT`]: https://hackage.haskell.org/package/mtl-2.2.2/docs/Control-Monad-Except.html#t:ExceptT
+
+However, we make the case that other languages which do have these explicit and
+reified exceptions as in Rust also use an exception vocabulary.
+Notably, Haskell calls the monad-transformer for adding exceptions [`ExceptT`].
+
+We also argue that even tho we are propagating exceptions manually,
+we are following tradition in that other languages have very different
+formulations of the exception idea.
+
+The benefit of familiarity, even if not a perfect match, as Niko puts it,
+helps in learning, particularly because Rust is not a language in lack of
+concepts to learn.
+
+[`try!`]: https://doc.rust-lang.org/nightly/std/macro.try.html
+
+## Breakage of the [`try!`] macro
+
+One possible result of introducing `try` as a keyword be that the old `try!`
+macro would break. This could potentially be avoided but with great technical
```

Comment:
Could you elaborate on (or maybe add a pointer to a previous discussion about) a strategy to avoid breaking `try!` and its challenges?

Reply:
A good question! This is mainly based on discussions with @Manishearth, so I refer you to them for a better answer.

Reply:
Breaks in the sense that you have to invoke it as `r#try` instead.